transport-classes-io_uring/src/main/java/io/netty/channel/uring/IoUringBufferRing.java

```diff
@@ -0,0 +1,413 @@
+/*
+ * Copyright 2025 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.uring;
+
+import io.netty.buffer.AbstractReferenceCountedByteBuf;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.internal.PlatformDependent;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import java.nio.channels.GatheringByteChannel;
+import java.nio.channels.ScatteringByteChannel;
+
+final class IoUringBufferRing {
+
+    private final long ioUringBufRingAddr;
+
+    private final short entries;
+
+    private final short bufferGroupId;
+
+    private final int ringFd;
+
+    private final ByteBuf[] userspaceBufferHolder;
+    private final int chunkSize;
+    private short nextIndex;
+    private boolean hasSpareBuffer;
+    private IoUringIoHandler source;
+
+    IoUringBufferRing(int ringFd, long ioUringBufRingAddr,
+                      short entries, short bufferGroupId,
+                      int chunkSize, IoUringIoHandler ioUringIoHandler
+    ) {
+        this.ioUringBufRingAddr = ioUringBufRingAddr;
+        this.entries = entries;
+        this.bufferGroupId = bufferGroupId;
+        this.ringFd = ringFd;
+        this.userspaceBufferHolder = new ByteBuf[entries];
+        this.nextIndex = 0;
+        this.chunkSize = chunkSize;
+        this.hasSpareBuffer = false;
+        this.source = ioUringIoHandler;
+    }
+
+    public void markReadFail() {
+        hasSpareBuffer = false;
+    }
+
+    public boolean hasSpareBuffer() {
+        return hasSpareBuffer;
+    }
+
+    void recycleBuffer(short bid) {
+        source.submitBeforeIO(new Runnable() {
```

Comment:
as all the possible `bid`s are known we should pre-create the `Runnable`s and reuse them to reduce GC pressure.